File: figmaclaw/in_context.py
"""Build composite "Usage in Context" frames in Figma DS component draft pages.

Orchestrates a sequence of use_figma calls that place a real product screen
next to a DS component set. Each call embeds pre-fetched section data (SVG or
PNG base64) within the 50K code string limit.

Architecture
------------
The Figma MCP plugin sandbox has no network access and no clientStorage.
The only data entry point is the use_figma code string (≤50,000 chars).

Strategy: split the source frame into sections, one use_figma call per section.
  - SVG preferred (live editable nodes): use if compressed SVG ≤ SVG_SIZE_LIMIT
  - PNG fallback (flat image fill): base64-encoded PNG @scale=0.25, always fits

Section positions are read from figma page frontmatter (frame_sections field,
written by the pull pass). No extra REST API call needed at build time.

See figmaclaw issues #35 and #38.
"""

from __future__ import annotations

import base64
import json
from dataclasses import dataclass
from pathlib import Path

from figmaclaw.figma_client import FigmaClient
from figmaclaw.figma_frontmatter import SectionNode

# Max data string length that safely fits inside a use_figma code call alongside helpers.
# Measured overhead per call: helpers (~9.8KB) + _find_page_js (~165 chars)
# + placeContextSection boilerplate (~175 chars) ≈ 10_200 chars total.
# Budget: 50_000 - 10_200 = 39_800. Use 38_000 for safety headroom.
SVG_SIZE_LIMIT = 38_000

# PNG export scale. 0.25 renders a 393×854 section at ~98×213px (~16KB PNG / ~21KB base64).
# Chosen to stay safely under SVG_SIZE_LIMIT for photo-heavy sections.
# (0.35 scale was too large: a 393×854px section produced a 32KB PNG / 43KB base64.)
PNG_SCALE = 0.25

# Plugin helpers JS — pasted at the top of every use_figma code string.
_HELPERS_PATH = Path(__file__).parent / "plugin" / "in-context.js"
_HELPERS_JS: str = _HELPERS_PATH.read_text()


@dataclass
class SectionData:
    """Pre-fetched data for one section of the source frame."""
    section: SectionNode
    kind: str    # 'svg' or 'png'
    data: str    # SVG markup or base64-encoded PNG bytes (no data: URI prefix)


async def fetch_section_data(
    client: FigmaClient,
    file_key: str,
    section: SectionNode,
) -> SectionData:
    """Fetch SVG or PNG for one section.

    Tries SVG first. Falls back to PNG @scale=0.25 if SVG exceeds SVG_SIZE_LIMIT.
    PNG always fits within the use_figma 50K code string limit at this scale.
    """
    svg_urls = await client.get_image_urls(file_key, [section.node_id], format="svg")
    svg_url = svg_urls.get(section.node_id)
    if svg_url:
        svg_bytes = await client.download_url(svg_url)
        svg_str = svg_bytes.decode("utf-8")
        if len(svg_str) <= SVG_SIZE_LIMIT:
            return SectionData(section=section, kind="svg", data=svg_str)

    # SVG too large — fall back to PNG
    png_urls = await client.get_image_urls(
        file_key, [section.node_id], format="png", scale=PNG_SCALE
    )
    png_url = png_urls.get(section.node_id)
    if not png_url:
        raise ValueError(
            f"No export URL returned for section {section.node_id!r} ({section.name!r})"
        )
    png_bytes = await client.download_url(png_url)
    b64 = base64.b64encode(png_bytes).decode("ascii")
    return SectionData(section=section, kind="png", data=b64)


def make_context_calls(
    *,
    target_file_key: str,
    target_page_id: str,
    container_name: str,
    frame_w: int,
    frame_h: int,
    comp_x: int,
    comp_y: int,
    comp_w: int,
    label: str,
    section_data_list: list[SectionData],
) -> list[dict[str, str]]:
    """Generate the ordered sequence of use_figma call specs.

    Returns a list of dicts: [{"file_key": ..., "description": ..., "code": ...}, ...]
    Each dict maps directly to the use_figma tool parameters.
    Execute them in order.

    Call 0:   createContextContainer — creates the outer frame
    Calls 1-N: placeContextSection — one per section
    Final:    addContextCaption — caption below the frame
    """
    x = comp_x + comp_w + 60
    y = comp_y

    calls: list[dict[str, str]] = []

    # Call 0: create container frame
    calls.append({
        "file_key": target_file_key,
        "description": f"Create context container frame '{container_name}' on page {target_page_id}",
        "code": _container_code(target_page_id, frame_w, frame_h, x, y, container_name),
    })

    # Calls 1-N: place each section
    for sd in section_data_list:
        calls.append({
            "file_key": target_file_key,
            "description": f"Place {sd.section.name!r} section ({sd.kind}) in '{container_name}'",
            "code": _section_code(target_page_id, container_name, sd),
        })

    # Final: add caption
    calls.append({
        "file_key": target_file_key,
        "description": f"Add caption to '{container_name}'",
        "code": _caption_code(target_page_id, container_name, label),
    })

    return calls


def _find_page_js(page_id: str) -> str:
    return f"""let page = null;
for (const p of figma.root.children) {{
  if (p.id === {json.dumps(page_id)}) {{ page = p; break; }}
}}
if (!page) throw new Error('Page {page_id} not found');
await figma.setCurrentPageAsync(page);"""


def _container_code(page_id: str, w: int, h: int, x: int, y: int, name: str) -> str:
    return f"""{_HELPERS_JS}

{_find_page_js(page_id)}

const frame = createContextContainer(page, {w}, {h}, {x}, {y}, {json.dumps(name)});
return frame.id;
"""


def _section_code(page_id: str, container_name: str, sd: SectionData) -> str:
    s = sd.section
    # Embed data as a JS template literal (backtick string).
    # SVG may contain backticks — escape them. PNG base64 never contains backticks.
    if sd.kind == "svg":
        data_js = "`" + sd.data.replace("\\", "\\\\").replace("`", "\\`") + "`"
    else:
        data_js = json.dumps(sd.data)  # plain string, no backtick issues

    return f"""{_HELPERS_JS}

{_find_page_js(page_id)}

const SECTION_DATA = {data_js};
await placeContextSection(page, {json.dumps(container_name)}, {{
  type: {json.dumps(sd.kind)},
  data: SECTION_DATA,
  x: {s.x}, y: {s.y}, w: {s.w}, h: {s.h},
  name: {json.dumps(s.name)},
}});
"""


def _caption_code(page_id: str, container_name: str, label: str) -> str:
    return f"""{_HELPERS_JS}

{_find_page_js(page_id)}

await figma.loadFontAsync({{ family: 'Inter', style: 'Regular' }});
await addContextCaption(page, {json.dumps(container_name)}, {json.dumps(label)});
"""

File: figmaclaw/plugin/in-context.js
// ── In-context usage section ───────────────────────────────────────────────────
// Builds a composite "Usage in Context" frame from individual sections of a real
// product screen. Each section is placed as either live SVG nodes or a PNG image
// fill, depending on its compressed size.
//
// ── Architecture (read this first) ────────────────────────────────────────────
//
// The MCP plugin sandbox has NO network access and NO clientStorage. The only
// way to get external data into the plugin is to embed it directly in the
// use_figma code string (50K char limit per call).
//
// Because a full screen SVG is ~2MB (too large), we split the screen into
// sections and make ONE use_figma call per section. Each call embeds that
// section's pre-fetched data (SVG string or base64 PNG).
//
// The orchestration (fetching, compressing, sequencing calls) is done by
// figmaclaw server-side Python. This file contains only the plugin-side helpers.
//
// ── Section decision rule ─────────────────────────────────────────────────────
//
//   For each section of the source frame:
//     1. Export as SVG via REST API → compress with svgo
//     2. If compressed SVG ≤ 38,000 chars → embed SVG → createNodeFromSvg
//        Result: real editable Figma vector nodes (pixel perfect, live)
//     3. Else → try raster (PNG then JPG) at scales [0.5, 0.35, 0.25] in order
//        Use first format+scale whose base64 fits under 38,000 chars
//        Result: flat raster image (not editable; JPG preferred for photos)
//   Overhead per call: ~10,200 chars (helpers + boilerplate), budget = 50,000 - 10,200.
//
//   Sections with complex fills / embedded avatar images are PNG candidates.
//   Simple sections (text + shapes only) compress well enough for SVG.
//
// ── Insights Tab section map (file 6nAmiusEvU31Z3fosx6vuo, frame 7424:15980) ──
//
//   Node ID      | Name             | y    | w×h      | Approach
//   -------------|------------------|------|----------|----------------
//   7424:16018   | Header           |    0 | 393×116  | SVG  (11.1KB)
//   7424:17814   | Top Questions    |  132 | 361×252  | PNG  (17.1KB b64)
//   7429:19265   | Top Comments     |  404 | 361×252  | PNG  (16.9KB b64)
//   7429:19382   | Top Reactions    |  676 | 361×252  | PNG  (17.5KB b64)
//   7429:19718   | Pinned Messages  |  948 | 361×252  | PNG  (11.7KB b64)
//   7429:19837   | Trending Topics  | 1220 | 361×198  | SVG  (39.5KB)
//   7429:104446  | Bottom bar       | 1434 | 393×116  | PNG   (3.7KB b64)
//
//   Frame total: 393×1584. Background color: #0D0D0D (dark).
//   Sections at x=16 are inset from the frame edge. Header and bottom bar at x=0.
//
// ── Call sequence (figmaclaw Python orchestrates) ─────────────────────────────
//
//   Call 0  — createContextContainer  → returns container frame node ID
//   Call 1  — placeContextSection (Header, SVG)
//   Call 2  — placeContextSection (Top Questions, PNG)
//   Call 3  — placeContextSection (Top Comments, PNG)
//   Call 4  — placeContextSection (Top Reactions, PNG)
//   Call 5  — placeContextSection (Pinned Messages, PNG)
//   Call 6  — placeContextSection (Trending Topics, SVG)
//   Call 7  — placeContextSection (Bottom bar, PNG)
//
//   Each call after 0 finds the container frame by name and appends to it.
//
// ── API key ───────────────────────────────────────────────────────────────────
//
//   Figma REST API calls happen server-side (figmaclaw Python), not here.
//   Read the API key from linear-git/.env → FIGMA_API_KEY at call time.
//   Never store or commit the key.
//
// ─────────────────────────────────────────────────────────────────────────────


// Call 0 — createContextContainer
// ─────────────────────────────────────────────────────────────────────────────
// Creates the outer frame that holds all sections. Call once before the section
// placement calls. Returns the frame node ID so subsequent calls can find it.
//
// Parameters:
//   page      — PageNode (must already be current page)
//   w, h      — source frame dimensions (393, 1584 for Insights Tab)
//   x, y      — canvas position (x = comp.x + comp.width + 60, y = comp.y)
//   name      — unique name used by subsequent calls to find this frame
//               e.g. 'ctx-insights-tab-7424-15980'
//   bgColor   — background fill color {r,g,b} (default dark: {r:0.05,g:0.05,b:0.05})
//
// Example:
//   const frame = createContextContainer(page, 393, 1584,
//     comp.x + comp.width + 60, comp.y,
//     'ctx-insights-tab-7424-15980');
//   return frame.id;  // pass to subsequent calls
//
function createContextContainer(page, w, h, x, y, name, bgColor) {
  const bg = bgColor ?? { r: 0.051, g: 0.051, b: 0.051 };
  const frame = figma.createFrame();
  frame.name = name;
  frame.resize(w, h);
  frame.x = x;
  frame.y = y;
  frame.fills = [{ type: 'SOLID', color: bg }];
  frame.clipsContent = true;
  frame.cornerRadius = 12;
  page.appendChild(frame);
  return frame;
}


// Calls 1–N — placeContextSection
// ─────────────────────────────────────────────────────────────────────────────
// Places one section inside the container frame. Call once per section.
// The container frame is looked up by name (containerName) on the current page.
//
// Parameters:
//   page          — PageNode (must already be current page)
//   containerName — name passed to createContextContainer
//   section       — object describing the section:
//     {
//       type:   'svg' | 'png' | 'jpg',
//       data:   string,   // SVG markup OR base64-encoded raster bytes
//       x:      number,   // x position relative to container frame (0 or 16)
//       y:      number,   // y position relative to container frame
//       w:      number,   // original section width in Figma units
//       h:      number,   // original section height in Figma units
//       name:   string,   // descriptive name e.g. 'Header', 'Trending Topics'
//     }
//
// Example (SVG section — figmaclaw injects the svgString at call time):
//   const SVG_DATA = `<svg ...>...</svg>`;  // injected by figmaclaw
//   await placeContextSection(page, 'ctx-insights-tab-7424-15980', {
//     type: 'svg', data: SVG_DATA,
//     x: 0, y: 0, w: 393, h: 116, name: 'Header'
//   });
//
// Example (PNG section — figmaclaw injects the base64 string at call time):
//   const PNG_B64 = 'iVBORw0KGgo...';  // injected by figmaclaw (base64, no data: prefix)
//   await placeContextSection(page, 'ctx-insights-tab-7424-15980', {
//     type: 'png', data: PNG_B64,
//     x: 16, y: 132, w: 361, h: 252, name: 'Top Questions'
//   });
//
async function placeContextSection(page, containerName, section) {
  // Find container
  const container = page.children.find(n => n.name === containerName && n.type === 'FRAME');
  if (!container) throw new Error(`Container frame '${containerName}' not found on current page`);

  let node;

  if (section.type === 'svg') {
    node = figma.createNodeFromSvg(section.data);
    node.name = section.name;
    node.x = section.x;
    node.y = section.y;
    // createNodeFromSvg returns a FRAME sized to the SVG viewport
    // Resize to match original Figma dimensions if they differ
    if (Math.abs(node.width - section.w) > 1 || Math.abs(node.height - section.h) > 1) {
      node.resize(section.w, section.h);
    }
    container.appendChild(node);

  } else {
    // PNG: decode base64 → create image → rect fill
    const bytes = figma.base64Decode(section.data);
    const img = figma.createImage(bytes);
    node = figma.createRectangle();
    node.name = section.name;
    node.resize(section.w, section.h);
    node.x = section.x;
    node.y = section.y;
    node.fills = [{ type: 'IMAGE', scaleMode: 'FILL', imageHash: img.hash }];
    container.appendChild(node);
  }

  return node.id;
}


// Optional: addContextCaption
// ─────────────────────────────────────────────────────────────────────────────
// Adds a small label below the container frame. Call after all sections placed.
//
// Example:
//   await figma.loadFontAsync({ family: 'Inter', style: 'Regular' });
//   addContextCaption(page, 'ctx-insights-tab-7424-15980',
//     'Mobile Insights Tab — Community in Live');
//
async function addContextCaption(page, containerName, labelText) {
  const container = page.children.find(n => n.name === containerName && n.type === 'FRAME');
  if (!container) return;

  await figma.loadFontAsync({ family: 'Inter', style: 'Regular' });
  const cap = figma.createText();
  cap.fontName = { family: 'Inter', style: 'Regular' };
  cap.fontSize = 10;
  cap.characters = labelText;
  cap.fills = [{ type: 'SOLID', color: { r: 0.45, g: 0.45, b: 0.45 } }];
  cap.x = container.x;
  cap.y = container.y + container.height + 8;
  page.appendChild(cap);
  return cap;
}
